### PR TITLE
Add lookup index mode

### PR DESCRIPTION
**Pull request by dnhatn**
This change introduces a new index mode, `lookup`, for indices intended for lookup operations in ES|QL. Lookup indices must have a single shard and be replicated to all data nodes by default. Aside from these requirements, they function as standard indices. Documentation will be added later when the lookup operator in ES|QL is implemented.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/IndexMode.java [hunk 1]
```diff
@@ -308,6 +309,87 @@ public SourceFieldMapper.Mode defaultSourceMode() {
         public String getDefaultCodec() {
             return CodecService.BEST_COMPRESSION_CODEC;
         }
+    },
+    LOOKUP("lookup") {
+        @Override
+        void validateWithOtherSettings(Map<Setting<?>, Object> settings) {
+            Integer defaultNumberOfShards = IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING.get(Settings.EMPTY);
+            Object providedNumberOfShards = settings.getOrDefault(IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING, defaultNumberOfShards);
+            if (Objects.equals(providedNumberOfShards, defaultNumberOfShards) == false) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/IndexMode.java:318]: Maybe just check whether `providedNumberOfShards` is 1?

File: server/src/main/java/org/elasticsearch/index/IndexMode.java [hunk 2]
```diff
@@ -308,6 +309,87 @@ public SourceFieldMapper.Mode defaultSourceMode() {
...
+                    "index with [lookup] mode must have [index.number_of_shards] set to 1 or unset; provided " + providedNumberOfShards
+                );
+            }
+        }
+
+        @Override
+        public Settings defaultIndexSettings() {
+            return Settings.builder()
+                .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                .put(IndexMetadata.SETTING_AUTO_EXPAND_REPLICAS, "0-all")
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/IndexMode.java:329]: Should we also validate whether auto expand replicas is really enabled? This just defines the defaults, but doesn't enforce that it is actually used?

File: server/src/main/java/org/elasticsearch/index/IndexMode.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/IndexMode.java]: ++, I pushed https://github.com/elastic/elasticsearch/pull/115143/commits/d84de0792401c5d0a209b8f3f27e12791a0d56ba

File: server/src/main/java/org/elasticsearch/index/IndexMode.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/IndexMode.java:328]: I may be missing something, but the default settings are copied to the index settings when creating an index: https://github.com/elastic/elasticsearch/pull/115143/files#diff-c02a1326ed501a89ce49b7e7a27df4bb9a2256785d00d7278bccb3609334b939R1112

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/IndexMode.java:328]: I see, the default settings are added after template and create index request settings have been applied. Maybe default settings is the wrong name? Since the settings from index mode don't act as default, but are enforced?

Additionally I think we need to mimic this logic in two other places too for consistency:

1. `MetadataIndexTemplateService#validateIndexTemplateV2(...)` to ensure that index mode default settings are applied during template validation.
2. `TransportSimulateIndexTemplateAction#resolveTemplate(...)` to ensure that index mode default settings are applied during simulate template api.

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/IndexMode.java:329]: `auto_expand_replicas` setting defaults to `0-all` and can be overridden by users, while the `number_of_shards` setting is enforced and must always be 1.

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/IndexMode.java:329]: Reading the PR description: `... Lookup indices must have a single shard and be replicated to all data nodes by default. ...`, is just apply a default for `auto_expand_replicas` to `0-all` sufficient?

**Review by nik9000 — APPROVED** [review]
I'm good with it but I'd wait on MVG to approve too. He knows a ton more about this stuff.

File: server/src/main/java/org/elasticsearch/index/IndexMode.java (hunk 2 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/IndexMode.java:329]: I spoke with Nhat about this and only `number_of_shards` must be 1. Auto expand replicas is optional and seen as an optimization for the enrich join.

File: server/src/main/java/org/elasticsearch/index/IndexMode.java [hunk 3]
```diff
@@ -468,6 +551,13 @@ public String getDefaultCodec() {
         return CodecService.DEFAULT_CODEC;
     }
 
+    /**
+     * The default index settings provided by this index mode.
+     */
+    public Settings defaultIndexSettings() {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/IndexMode.java:557]: Like we discussed yesterday. I think instead of adding defaultIndexSettings() method to index mode and then invoking that from `MetadataCreateIndexService`, `validateWithOtherSettings` and `MetadataIndexTemplateService` (currently missing in this change), we can implement this via an index settings provider (`IndexSettingProvider`)? If index mode is lookup, then it would return a settings instance with `index.number_of_shards` and `index.auto_expand_replicas` settings set to `1` and `0-all` respectively.  The validation in `LOOKUP.validateWithOtherSettings(...)` should then remain.

**Review by martijnvg — COMMENTED** [review]
I left one comment about an alternative for adding `IndexMode#defaultIndexSettings()`.

**Comment by dnhatn** [comment]
Thanks, Martijn, for the suggestion! Using the index setting provider looks cleaner. Would you mind taking another look?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 4]
```diff
@@ -308,7 +308,12 @@ private void onlyCreateIndex(
         final CreateIndexClusterStateUpdateRequest request,
         final ActionListener<AcknowledgedResponse> listener
     ) {
-        normalizeRequestSetting(request);
+        try {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:311]: Is this change still required?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 5]
```diff
@@ -1599,6 +1604,15 @@ static IndexMetadata validateResize(
             // of if the source shards are divisible by the number of target shards
             IndexMetadata.getRoutingFactor(sourceMetadata.getNumberOfShards(), INDEX_NUMBER_OF_SHARDS_SETTING.get(targetIndexSettings));
         }
+        if (targetIndexSettings.hasValue(IndexSettings.MODE.getKey())) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:1607]: This also helps when someone changes to and from logsdb and tsdb 👍 .

**Review by martijnvg — APPROVED** [review]
Left one question, LGTM otherwise.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java (hunk 4 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:311]: Yes, since validation can throw exceptions, and the caller does not expect them from this method.

**Comment by dnhatn** [comment]
Thanks @martijnvg.
